Frameworks/CoreGraphics/CGContextCairo.mm

```diff
@@ -1887,17 +1898,48 @@
     CGAffineTransform userTransform =
         CGAffineTransformConcat(curState->curTransform, CGAffineTransformMake(1.0f / _scale, 0, 0, -1.0f / _scale, 0, height / _scale));
 
-    // Apply the text position in user space, not in text space
-    userTransform = CGAffineTransformTranslate(userTransform, curState->curTextPosition.x, curState->curTextPosition.y);
+    CGAffineTransform finalTransform = CGAffineTransformConcat(textTransform, userTransform);
 
-    // Apply the two transforms giving us the final result
-    CGAffineTransform transform = CGAffineTransformConcat(textTransform, userTransform);
-    imgRenderTarget->SetTransform(D2D1::Matrix3x2F(transform.a, transform.b, transform.c, transform.d, transform.tx, transform.ty));
+    if ((fabs(finalTransform.a * glyphRun->fontEmSize) <= c_glyphThreshold &&
+         fabs(finalTransform.b * glyphRun->fontEmSize) <= c_glyphThreshold) ||
+        (fabs(finalTransform.c * glyphRun->fontEmSize) <= c_glyphThreshold &&
+         fabs(finalTransform.d * glyphRun->fontEmSize) <= c_glyphThreshold)) {
+        TraceWarning(TAG, L"Glyphs too small to be rendered");
+        return;
+    }
 
+    // Set the brush color to the current values from the context.
+    D2D1::ColorF brushColor =
+        D2D1::ColorF(curState->curFillColor.r, curState->curFillColor.g, curState->curFillColor.b, curState->curFillColor.a);
+
+    imgRenderTarget->BeginDraw();
     // Draw the glyph using ID2D1RenderTarget
     ComPtr<ID2D1SolidColorBrush> brush;
     THROW_IF_FAILED(imgRenderTarget->CreateSolidColorBrush(brushColor, &brush));
-    imgRenderTarget->DrawGlyphRun(D2D1::Point2F(0, 0), glyphRun, brush.Get(), DWRITE_MEASURING_MODE_NATURAL);
+    if (textTransform.a == 1.0f && fabs(textTransform.d) == 1.0f && textTransform.b == 0.0f && textTransform.c == 0.0f) {
```

Comment:
nit: should probably fabs(textTransform.a) too? horizontal flips seem like they can also use the fast path